pytensor/scan/rewriting.py

```diff
@@ -1103,6 +1111,71 @@ def sanitize(x):
         return at.as_tensor_variable(x)
 
 
+@node_rewriter([Scan])
+def while_scan_merge_subtensor_last_element(fgraph, scan_node):
+    """
+    Replace while_scan_out[abs(min(tap)):][-1] by while_scan_out[-1], for
+    recurring outputs, asserting that at least one step occurs.
+    Only the first step can be ensured by the inputs alone (i.e., `n_steps > 0`),
+    as the while scan could abort earlier anytime after that. This means it is
+    not possible to replace while_scan_out[abs(min(tap)):][-i]
+    by while_scan_out[-i], for -i != -1.
+    """
+    op = scan_node.op
+
+    if not op.info.as_while:
+        return None
+
+    # Optimization is not implemented form mit-mot
+    recurrent_outputs = op.outer_mitsot_outs(scan_node.outputs) + op.outer_sitsot_outs(
+        scan_node.outputs
+    )
+    recurrent_outputs_taps_slices = (
+        op.info.mit_sot_in_slices + op.info.sit_sot_in_slices
+    )
+
+    n_steps = scan_node.inputs[0]
+    non_zero_steps_cond = n_steps > 0
+    assert_non_zero_steps_op = Assert("n_steps > 0")
+
+    subtensor_merge_replacements = {}
+
+    # Iterate over all nodes that are two computations below the while scan
+    for node in get_clients_at_depth(fgraph, scan_node, depth=2):
+        if not isinstance(node.op, Subtensor):
+            continue
+
+        u = node.inputs[0]
+        if not (u.owner and isinstance(u.owner.op, Subtensor)):
+            continue
+
+        x = u.owner.inputs[0]
+        if x not in recurrent_outputs:
+            continue
+
+        slice1 = get_idx_list(u.owner.inputs, u.owner.op.idx_list)
+        slice2 = get_idx_list(node.inputs, node.op.idx_list)
+
+        min_tap = abs(min(recurrent_outputs_taps_slices[recurrent_outputs.index(x)]))
```

Comment:
Yes, although I think the `abs` is a bit more readable?

Reply:
It is true that with `abs`, the sign of `min_tap` is immediately obvious. However, this comes at the cost of making the sign of the input of `abs` not immediately obvious, because `abs` is irreversible and throws away one bit of information.
Besides, it can be argued that using `abs` implicitly suggests that the input can be of either sign… I did ask myself the question when reading because I was not sure what taps were.
That's why I think that a negation, with an assert about the sign of `min_tap`, is the best of both worlds at better conveying the meaning and at making the sign obvious.
But that's also a detail, it is fine with me if you don't feel like it.

Reply:
I gave it another thought, and I think `abs` is still better, because it makes it clear that the start of the slice is positive `x[start: stop: step]` in a way that is not as clear if we flip the minimum tap. I find that more important to gauge than re-emphasizing that taps are always negative.